lib/app.dart

```diff
@@ -28,51 +28,131 @@ library;
-/// does not need to manage any state itself. The state is managed through
-/// riverpod and so it is a [ConsumerWidget].
-
-class RattleApp extends ConsumerWidget {
+/// This widget manages the application's lifecycle and handles cleanup
+/// operations when the app is about to close. It uses [ConsumerStatefulWidget]
+/// to interact with Riverpod providers and [WindowListener] to handle
+/// window-related events, particularly for desktop platforms.
+class RattleApp extends ConsumerStatefulWidget {
   const RattleApp({super.key});
 
-  /// Build the root widget as a [MaterialApp] widget, setting up the app theme,
-  /// and populating the widget with the Rattle home page widget.
+  @override
+  ConsumerState<RattleApp> createState() => _RattleAppState();
+}
 
+class _RattleAppState extends ConsumerState<RattleApp> with WindowListener {
+  /// Initializes the state and sets up window management.
   @override
-  Widget build(BuildContext context, WidgetRef ref) {
-    // Initialise the R process.
+  void initState() {
+    super.initState();
+    windowManager.addListener(this);
+    _init();
+  }
 
-    // 20240809 On Windows this does not get run - that is main.R is not in the
-    // Console. Delaying until feature/dataset/popup.dart seems to work. Moving
-    // main.R into dataset_prep.R delays the setup too much.
-    //
-    // 20240810 Revert to this for now and try to solve the Windows issue. Is it
-    // because my Windows are too slow and this starts before the Console is
-    // running?
+  /// Removes this object as a window listener when the widget is disposed.
+  @override
+  void dispose() {
+    windowManager.removeListener(this);
+    super.dispose();
+  }
 
-    rStart(context, ref);
+  /// Initializes window management settings.
+  void _init() async {
+    // Prevent the window from closing by default
+    await windowManager.setPreventClose(true);
+    setState(() {});
+  }
+
+  /// Handles the window close event.
+  ///
+  /// This method is called when the user attempts to close the window.
+  /// It shows a confirmation dialog and performs cleanup if the user confirms.
+  @override
+  void onWindowClose() async {
+    bool isPreventClose = await windowManager.isPreventClose();
+    if (isPreventClose) {
+      _showCloseConfirmationDialog();
+    }
+  }
+
+  void _showCloseConfirmationDialog() {
+    showDialog(
+      context: context,
+      builder: (BuildContext context) {
+        return AlertDialog(
+          title: const Text('Are you sure you want to close Rattle?'),
+          content: const Text('Any unsaved work will be lost.'),
+          actions: [
+            TextButton(
+              child: const Text('Cancel'),
+              onPressed: () => Navigator.of(context).pop(false),
+            ),
+            TextButton(
+              child: const Text('Close'),
+              onPressed: () => {
+                cleanUpTempDirs(),
```

Comment:
Also add the method `_saveScript`

Reply:
resolved in [4271f34](https://github.com/gjwgit/rattleng/pull/442/commits/4271f34746df803cd749f7e255728829686ca6da)